DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/mmhost/README.md

```diff
@@ -0,0 +1,335 @@
+
+
+## Purpose
+When you design an IP component for an FPGA system, that system will often dictate the interface requirements of your IP component. This tutorial shows how to use `annotated_ptr` to configure Avalon memory-mapped host data interfaces. An Avalon memory-mapped host interface allows your IP component to send read or write requests to one or more Avalon memory-mapped agent interfaces. To learn more about Avalon memory-mapped host interfaces and Avalon memory-mapped agent interfaces, please refer to the appropriate section of the [Avalon Interface Specifications](https://www.intel.com/content/www/us/en/docs/programmable/683091/22-3/memory-mapped-interfaces.html). 
+
+![](assets/basic_avhost.svg)
+
+The compiler will infer Avalon memory-mapped host interfaces for your design when your kernel includes one or more pointer arguments. As with scalar kernel arguments, pointer arguments can be passed to your kernel via a `conduit` interface or your component's control/status register (CSR). By default, pointer arguments will be passed to your IP component through the CSR. For more details on kernel arguments, please see `<Temporary: https://github.com/intel-sandbox/wenkaixu.ip-auth-demos/tree/2024_0/mm-host>`. By default, the Intel® oneAPI DPC++/C++ Compiler will produce a kernel with a single Avalon memory-mapped host interface that will be shared amongst those pointers. 
+
+#### Example 1: A kernel with multiple pointer arguments
+```c++
+struct PointerIP{
+  int *x; 
+  int *y; 
+  int *z;
+  int size;
+
+  void operator()() const {
+    for (int i = 0; i < size; ++i) {
+      int mul = x[i] * y[i];
+      int add = mul + z[i];
+      y[i] = mul;
+      z[i] = add;
+    }
+  }
+};
+```
+![](assets/visio-naive.svg)
+
+### Using `annotated_ptr` to configure Avalon memory-mapped host interfaces
+
+You can override the default behaviour of a pointer argument by declaring an `annotated_ptr` kernel argument instead.
+
+#### Example 2: A kernel with a single customized Avalon memory-mapped host interface
+```c++
+struct SingleMMIP {
+  annotated_ptr<int, decltype(properties{
+    register_map,
+    buffer_location<kBL1>,
+    awidth<32>, 
+    dwidth<32>, 
+    latency<0>, 
+    read_write_mode<read>,
+    maxburst<4>, 
+    wait_request<true>
+  })> my_pointer; 
+  void operator()() const {
+    ...
+    // Kernel code
+    my_pointer[x] = y;
+    ...
+  }
+}
+```
+
+As with scalar kernel arguments and the `annotated_ptr` type, you can use the following parameters to control how the pointer is passed to your kernel:s
```

Comment:
This sentence needs some reworking.

Reply:
Yes, what are you trying to say here Hakim?

Reply:
Maybe just say: The following table describes additional properties that you can use to customize the interface.

For stable, may refer to hls docs and specify why a customer may want to do this. Paul can help with that too.